add support for shape note stafftypes

CraigFisher opened the pull request:
This work includes the removal of the "StaffTypes Enum".  My reasoning for this is as follows:

First, only three members of the StaffTypes enum were being explicitly used in the code (Traditional, Default Tab, Default Percussion).

Second, these three members could easily be replaced with the existing "Staffgroups" enum combined with the existing "default staffgroup" function.  (So StaffTypes was replicating existing code).

Third, the names of all hardcoded stafftypes all already documented through the "_presets" array.

Finally, the StaffTypes enum used to determine which stafftypes were available to the various spinners and other selection lists.  However, in my coming project (after this pull request), the user will be able to dynamically load new stafftypes, and this means that the available stafftypes must be able to change dynamically.  However, since enums cannot dynamically change, so it made sense to eliminate "StaffTypes" altogether.

lasconic commented on libmscore/stafftype.cpp:
```diff
@@ -150,11 +273,16 @@ bool StaffType::isSameStructure(const StaffType& st) const
             return false;
 
-      if (_group != StaffGroup::TAB) {                      // common to pitched and percussion
+      if (_group == StaffGroup::TAB) {                      // common to pitched and percussion
```
I just had a quick glance but this seems weird.

CraigFisher replied:
let me check that myself.  that DOES seem weird

lasconic commented:
@mgavioli Hello, this is still a work in progress but I'm curious about your opinion about the removal of the "StaffTypes Enum"

mgavioli commented:
It is hard to understand all the implications from diff files.

On the spot, I would believe that the current `StaffTypes` enum is required to properly deal with staff type presets. Note that, among other details, presets can be references from instruments.xml to assign a specific staff type to a specific instrument.

If the name seems not proper, it could be renamed into `BuiltInStaffTypePresets` or anything else.

mgavioli commented:
UPDATE: some archaeological research have ungrounded a few more evidences. That `enum` as I wrote it was a simple, unnamed, uninteresting, old-fashion, C `enum` ; it got its current name when all `enum`'s have been converted to `enum class <name> : <type>` constructs.

And its goal was to list all the **built-in** presets, those the programme can be sure they exist and which can be referenced from elsewhere (user presets have never been implemented, but the possibility of non-built-in presets was foreseen since the beginning).

So, I still believe that that `enum` is needed, in this or in some other form (name, type or whatever).

If staff configurability is going to be extended (for shape notes or anything else) and more built-in presets make sense, the `enum`can obviously be extended.

CraigFisher commented:
> And its goal was to list all the built-in presets, those the programme can be sure they exist and which can be referenced from elsewhere (user presets have never been implemented, but the possibility of non-built-in presets was foreseen since the beginning).

I understand that purpose.  However, if you run a search of "StaffTypes" in the entire code base, the only enums actually used are these:  

```
STANDARD, 
PERC_DEFAULT = StaffTypes::PERC_5LINE,
TAB_DEFAULT = StaffTypes::TAB_6COMMON
```

As you can see, two of them are just aliases (so we can get the default staffType).  However, we already have a method for getting the default StaffType using just the "StaffGroup": 

    const StaffType* StaffType::getDefaultPreset(StaffGroup grp)

When I replaced the StaffTypes enum, I didn't have to add anything to the code base to do it.  I just used other methods that already exist.

CraigFisher commented:
> If staff configurability is going to be extended (for shape notes or anything else) and more built-in presets make sense, the enumcan obviously be extended.

I wouldn't mind extending the enum, but there's one danger down the road.  In my Senior Project, which I'll be submitting soon, I've added the ability to configure and save new staff types (to allow for a wide variety of alternative notation schemes).

This also means the user could *dynamically* add more stafftypes to the presets.  If we allow dynamically-generated stafftypes AND keep the StaffTypes enum, we'll have to range checks, which I think are dangerous.  Let me explain:

In the following code, we're using the StaffTypes enum to "index in" to the correct stafftype preset.

```
const StaffType* StaffListItem::staffType() const
      {
      return StaffType::preset(StaffTypes((staffTypeIdx())));
      }
```

Now imagine the user has added one more stafftype to their presets.  Then, if the user tries to select their dynamically-loaded stafftype, the new index will be out of range in the StaffTypes enum.

The simple solution is to ignore the StaffTypes enum and use only indexes here.  This is what I've done in my code.  I've changed the StaffType::preset method to take an integer instead of a StaffType, and I ignore the enum.

```
const StaffType* StaffListItem::staffType() const
      {
      return StaffType::preset((staffTypeIdx()));
      }
```

However, if we keep using the enum, we'd have to do a range check every time the StaffTypes enum is queried:

```
const StaffType* StaffListItem::staffType() const
      {
      if (staffTypeIdx() < size_of(StaffTypes) -1)
           return StaffType::preset(StaffTypes((staffTypeIdx())));
      else
            // CHECK THE DYNAMIC STAFFTYPES
      }
```

While we CAN do this, if someone ever forgets to do the range check, the code will break (at least for custom stafftypes).  Thus, I feel it's more *maintainable* to remove the StaffTypes enum altogether so that no one's ever tempted to forget the range check.

mgavioli commented:
Custom staff presets (which would come handy also for tabs, for instance) was exactly my point: there are cases -- presets referenced in `instruments.xml` is one, but there might be other -- where the **built-in** presets (those we are sure they exist) are referenced. That `enum` was intended for this.

I assume that, if custom presets are implemented, another, different array will keep the list of presets **actually present** in the score (or in a user dictionary, whatever), including built-in and custom presets. The code will rely on one or the other, according to the specific use case.

The name of the current `StaffTypes enum` might be misleading as well as it being referenced from other spots: the name and those references have been added after the fact, during some large-scale, bulk update of coding conventions and might be dubious (if you check the `stafftype.h` file history, you will discover that originally several of the spots you changed back to `(int)` indices were already them!...)

A rename of the `enum` -- to make its meaning clearer -- and a check of its references are probably needed and you probably already did (most of?) it. But I have the feeling that it is logically (and presumably functionally) needed.

CraigFisher commented:
I still don't understand why the enum is necessary.  But I'd be willing to stick it back in (possibly renamed) if you feel it's necessary.

I feel it's quite important not to actually use it for functions that will reference the entire list of presets:

```
const StaffType* StaffListItem::staffType() const
      {
      return StaffType::preset(StaffTypes((staffTypeIdx())));
      }
```

Of course, for functions related to xml loading, I think it's just as semantic to use the "default stafftype for a given StaffGroup".  i.e.

    staff->setStaffType(StaffType::getDefaultPreset(StaffGroup::PERCUSSION));

Let me know how I should proceed.  I'm not sure how long these pull requests usually take to get accepted, but I'd like to move it along if possible.

mgavioli commented:
OK. I'll try to put it as simple as I can: I think it is wise to keep the list of **built-in** presets distinct from the list of **all** presets (built-in and custom); the current `enum` -- possibly renamed -- seems fine to me, but another mean may be fine as well.

As an example, you may look at: https://github.com/musescore/MuseScore/blob/master/share/instruments/instruments.xml#L8480 where a specific preset is referenced as default preset for a specific instrument (a preset which is **not** the default for that staff group) and such a preset has to exist.

I can imagine that some kind of resetting to  factory defaults may also need such a list; and so on.

As for the code example you quote and other similar, I agree 100%: in fact, many of your changes restore the code as I originally write it; look for instance at `stafftype.h` as it was before the `enum` mass change in May 2014: https://github.com/musescore/MuseScore/blob/2f85f8e007c57acba1ce468576823a26c958e296/libmscore/stafftype.h#L245

About the time it takes for a PR to get merged, it may vary from a few hours to months...

CraigFisher commented:
Should I put the enum back in, then?

Also, how should this method figure in?

```
const StaffType* presetFromXmlName(QString& xmlName)
```

lasconic commented:
Ok. I finally tried the PR. I think we are trying too much in one go.. 
I would
1/ Not necessarily create built-in preset for 4 and 7 shape notes. (it would remove the clutter in the instrument wizard, if we have custom staff type in the future, people would be able to create them if they wish)
2/ Add controls in the standard staff config to deal with shape notes style
3/ Focus first on shape notation and then on custom staff type (or vice versa...)

Regarding the other changes.
I agree with Maurizio when he says *I think it is wise to keep the list of **built-in** presets distinct from the list of **all** presets (built-in and custom);*. In any case, it shouldn't be possible for users to delete or modify a built-in preset. I'm not sure how you plan to handle this in your future PR but maybe the whole change belong to this new PR and this one should be limited to the addition of the shape note style

CraigFisher commented:
> 1/ Not necessarily create built-in preset for 4 and 7 shape notes. (it would remove the clutter in the instrument wizard, if we have custom staff type in the future, people would be able to create them if they wish)

Yes, I agree that adding "Shape Note 4" and "Shape Note 7" directly to the default stafftype list seems awkward.  Of course, for a dedicated Shape Note user, whose main use of MuseScore *is* Shape Notes, it makes perfect sense to see it upfront.

What if the user could add Shape Notes to the dialog if they want them?

lasconic commented:
In my view, it's what should happen if we have a way to add more custom presets indeed.

CraigFisher commented:
> I agree with Maurizio when he says I think it is wise to keep the list of *built-in** presets distinct from the list of all presets (built-in and custom);*. In any case, it shouldn't be possible for users to delete or modify a built-in preset. 

I wouldn't want the user to modify prebuilt templates either.  I've thought about this months ago, and here's how I handle it in my current code:

```
static std::vector<StaffType> _prebuiltTemplates; //JUST THE BUILT IN TEMPLATES
static std::vector<StaffType> _userTemplates;     //JUST THE USER TEMPLATES
static std::vector<StaffType> _presets;           //COMBINED LIST
```

Only the `_userTemplates` vector is ever modified directly.  Then to merge the two vectors, I have this method:

```
static void copyUserTemplatesToPresets(std::vector<StaffType>&);
```

It's worth noting that there's nothing special about the "_presets" in the current code base that keeps it from being modified.  It's not a constant structure after all.  We just choose not to modify it.

CraigFisher commented:
> In my view, it's what should happen if we have a way to add more custom presets indeed.

This is definitely the path of my final project.  Of course, in my final project, I have a dedicated dialog where custom StaffTypes can be added, solving the problem nicely.  For now though, I don't have such a convenience.

lasconic commented:
> This is definitely the path of my final project
Then maybe it's easier to get the full picture first on the whole user preset thing and have this separated from the shape notes thing.

lasconic commented:
MuseScore master now have the concept of notehead scheme and support shape notes, but also notehead with pitch names and several other schemes.

anatoly-os commented:
In current master the use of `StaffTypes` is widely spread. A lot of changes have been made, so I'm not sure this PR still makes sense.
